### PR TITLE
timelock optimization

### PR DESCRIPTION
**Pull request by Flydexo**
Fixes: #19 

Hey @moodysalem, Sorry for the delay, tell me if that corresponds to what you expect, if so I will complete this PR with all the files with the same method

### CODE REVIEW

**Review by moodysalem — CHANGES_REQUESTED** [review]
not exactly, there is a StorePacking trait you can implement to keep the structs well named and pack them into a felt252

3 timestamps should fit in 192 bits, so you should use a felt252

File: src/timelock.cairo [hunk 1]
```diff
@@ -14,23 +18,119 @@ trait ITimelock<TStorage> {
...
 
+
+const U64_MASK: u128 = 0xFFFFFFFFFFFFFFFF;
+const TWO_POW_64: u128 = 0x10000000000000000;
+
+impl ThreeU64TupleStorePacking of StorePacking<(u64, u64, u64), felt252> {
+    fn pack(value: (u64, u64, u64)) -> felt252 {
+        let (a, b, c) = value;
+        u256 {
+            low: a.into()+b.into()*TWO_POW_64,
```
Comment (moodysalem) [inline src/timelock.cairo:43]: you can use the two impl here

File: src/timelock.cairo [hunk 2]
```diff
@@ -14,23 +18,119 @@ trait ITimelock<TStorage> {
...
+            high: c.into()
+        }.try_into().unwrap()
+    }
+    fn unpack(value: felt252) -> (u64, u64, u64) {
+        let u256_value: u256 = value.into();
+        (
+            (u256_value.low & U64_MASK).try_into().unwrap(),
```
Comment (moodysalem) [inline src/timelock.cairo:50]: dividing and bit-anding separately is more expensive than just dividing once and using both the remainder quotient

Comment (moodysalem) [reply src/timelock.cairo:50]: also you should use the two impl here to unpack the low

File: src/timelock.cairo [hunk 3]
```diff
@@ -14,23 +18,119 @@ trait ITimelock<TStorage> {
...
     // Configure the delay and the window for call execution. This must be self-called via #queue.
-    fn configure(ref self: TStorage, delay: u64, window: u64);
+    fn configure(ref self: TStorage, delay_and_window: TimelockConfig);
 }
 
+
+const U64_MASK: u128 = 0xFFFFFFFFFFFFFFFF;
+const TWO_POW_64: u128 = 0x10000000000000000;
+
+impl ThreeU64TupleStorePacking of StorePacking<(u64, u64, u64), felt252> {
```
Comment (moodysalem) [inline src/timelock.cairo:39]: do these both belong in the timelock file?

File: src/timelock.cairo [hunk 4]
```diff
@@ -14,23 +18,119 @@ trait ITimelock<TStorage> {
...
 
     // Returns the delay and the window for call execution
-    fn get_configuration(self: @TStorage) -> (u64, u64);
+    fn get_configuration(self: @TStorage) -> TimelockConfig;
 
     // Transfer ownership, i.e. the address that can queue and cancel calls. This must be self-called via #queue.
     fn transfer(ref self: TStorage, to: ContractAddress);
     // Configure the delay and the window for call execution. This must be self-called via #queue.
-    fn configure(ref self: TStorage, delay: u64, window: u64);
+    fn configure(ref self: TStorage, delay_and_window: TimelockConfig);
```
Comment (moodysalem) [inline src/timelock.cairo:32]: rename to config

File: src/timelock.cairo [hunk 5]
```diff
@@ -14,23 +18,119 @@ trait ITimelock<TStorage> {
...
+
+#[derive(Copy, Drop, Serde)]
+struct ExecutionState {
+    started: u64,
+    executed: u64,
+    canceled: u64
+}
+
+#[inline(always)]
+impl ExecutionStateStorePacking of StorePacking<ExecutionState, (u64, u64, u64)> {
```
Comment (moodysalem) [inline src/timelock.cairo:78]: i'm pretty sure this doesn't work recursively, although i'm not 100% confident. meaning you need to call store packing in the impl

you can write a unit test that actually stores and reads the object

File: src/timelock.cairo [hunk 6]
```diff
@@ -69,18 +169,15 @@ mod Timelock {
...
+        execution_state: LegacyMap<felt252, ExecutionState>,
     }
 
     #[constructor]
-    fn constructor(ref self: ContractState, owner: ContractAddress, delay: u64, window: u64) {
+    fn constructor(ref self: ContractState, owner: ContractAddress, delay_and_window: TimelockConfig) {
         self.owner.write(owner);
-        self.delay.write(delay);
-        self.window.write(window);
+        self.delay_and_window.write(delay_and_window);
```
Comment (moodysalem) [inline src/timelock.cairo:180]: why not name this config

File: src/timelock.cairo [hunk 7]
```diff
@@ -113,10 +210,11 @@ mod Timelock {
             self.check_owner();
 
             let id = to_id(calls);
+            let execution_state = self.execution_state.read(id);
 
-            assert(self.execution_started.read(id).is_zero(), 'ALREADY_QUEUED');
+            assert(execution_state.started.is_zero(), 'ALREADY_QUEUED');
 
-            self.execution_started.write(id, get_block_timestamp());
+            self.execution_state.write(id, ExecutionState{started: get_block_timestamp(), executed: execution_state.executed, canceled: execution_state.canceled});
```
Comment (moodysalem) [inline src/timelock.cairo:217]: i think i woudl prefer just writing 0 for these fields

also looks like we're missing a check for canceled== 0?

it probably shouldn't be allowed to be re-queued if already canceled

**Review by moodysalem — CHANGES_REQUESTED** [review]
you might want to add a pre-commit hook for `scarb fmt`

File: src/timelock.cairo (hunk 3 above)
Comment (Flydexo) [reply src/timelock.cairo:39]: I can create a file dedicated to the timestamp optimization

File: src/timelock.cairo [hunk 8]
```diff
@@ -14,23 +18,108 @@ trait ITimelock<TStorage> {
...
+    }
+}
+
+#[derive(Copy, Drop, Serde)]
+struct ExecutionWindow {
+    earliest: u64,
+    latest: u64
+}
+
+#[inline(always)]
```
Comment (moodysalem) [inline src/timelock.cairo:104]: does inline work on traits?

File: src/timelock.cairo [hunk 9]
```diff
@@ -113,10 +199,16 @@ mod Timelock {
             self.check_owner();
 
             let id = to_id(calls);
+            let execution_state = self.execution_state.read(id);
 
-            assert(self.execution_started.read(id).is_zero(), 'ALREADY_QUEUED');
+            assert(execution_state.started.is_zero(), 'ALREADY_QUEUED');
+            assert(execution_state.canceled.is_non_zero(), 'PREVIOUSLY_CANCELED');
```
Comment (moodysalem) [inline src/timelock.cairo:205]: ty for adding this!

File: src/timelock.cairo [hunk 10]
```diff
@@ -160,26 +273,27 @@ mod Timelock {
...
+            let configuration = (self.get_configuration());
+
+            let earliest = start_time + configuration.delay;
 
-            let earliest = start_time + delay;
-            let latest = earliest + window;
+            let latest = earliest + configuration.window;
 
-            (earliest, latest)
+            ExecutionWindow { earliest: earliest, latest: latest }
```
Comment (moodysalem) [inline src/timelock.cairo:288]: you don't have to repeat the value if the variable name is the same as the key

File: src/timelock.cairo [hunk 11]
```diff
@@ -14,23 +18,108 @@ trait ITimelock<TStorage> {
...
+}
+
+#[derive(Copy, Drop, Serde)]
+struct ExecutionWindow {
+    earliest: u64,
+    latest: u64
+}
+
+#[inline(always)]
+impl ExecutionWindowStorePacking of StorePacking<ExecutionWindow, u128> {
```
Comment (moodysalem) [inline src/timelock.cairo:105]: we don't actually store this, you can remove the store impl

File: src/timelock.cairo [hunk 12]
```diff
@@ -160,26 +273,27 @@ mod Timelock {
             results
         }
 
-        fn get_execution_window(self: @ContractState, id: felt252) -> (u64, u64) {
-            let start_time = self.execution_started.read(id);
+        fn get_execution_window(self: @ContractState, id: felt252) -> ExecutionWindow {
+            let start_time = self.execution_state.read(id).started;
 
             // this is how we prevent the 0 timestamp from being considered valid
             assert(start_time != 0, 'DOES_NOT_EXIST');
```
Comment (moodysalem) [inline src/timelock.cairo:280]: hmmm maybe we could just check that canceled is nonzero as well and remove the other logic changes. sorry, i realize this is out of scope for the initial pr.

File: src/timelock.cairo [hunk 13]
```diff
@@ -125,26 +177,47 @@ mod Timelock {
...
         fn execute(ref self: ContractState, mut calls: Span<Call>) -> Array<Span<felt252>> {
             let id = to_id(calls);
 
-            assert(self.executed.read(id).is_zero(), 'ALREADY_EXECUTED');
+            let execution_state = self.execution_state.read(id);
 
-            let (earliest, latest) = self.get_execution_window(id);
+            assert(execution_state.executed.is_zero(), 'ALREADY_EXECUTED');
+
+            let execution_window = self.get_execution_window(id);
```
Comment (moodysalem) [inline src/timelock.cairo:205]: this causes a second read of the same slot

not super costly, but wasteful
better to revert if already executed or canceled from this function, since there is no execution window

**Review by moodysalem — APPROVED** [review]

**Comment by moodysalem** [comment]
if you resolve conflicts we can merge
